sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftIO.java

```diff
@@ -0,0 +1,708 @@
+     * href="http://docs.oracle.com/javase/tutorial/essential/io/find.html" >Java Filesystem glob
+     * patterns</a> ("*", "?", "[..]") are supported.
+     */
+    public Read from(String filePattern) {
+      return from(StaticValueProvider.of(filePattern));
+    }
+
+    /** Same as {@code from(filepattern)}, but accepting a {@link ValueProvider}. */
+    public Read from(ValueProvider<String> filepattern) {
+      return toBuilder().setFilePattern(filepattern).build();
+    }
+
+    /**
+     * Returns a transform for reading Thrift files that decompresses all input files using the
+     * specified compression type.
+     *
+     * <p>If no compression type is specified, the default is {@link Compression#AUTO}. In this
+     * mode, the compression type of the file is determined by it's extension via {@link
+     * Compression#detect(String)}.
+     */
+    public Read withCompression(Compression compression) {
+      return toBuilder().setCompression(compression).build();
+    }
+
+    @Override
+    public PCollection<Document> expand(PBegin input) {
+
+      return input
+          .apply("Create filepattern", Create.ofProvider(getFilePattern(), StringUtf8Coder.of()))
+          .apply(FileIO.matchAll())
+          .apply(FileIO.readMatches())
+          .apply(readFiles());
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.add(
+          DisplayData.item("filePattern", getFilePattern()).withLabel("Input File Pattern"));
+      builder.add(
+          DisplayData.item("compressionType", getCompression().toString())
+              .withLabel("Compression Type"));
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+
+  /** Creates a {@link Sink} for use with {@link FileIO#write} and {@link FileIO#writeDynamic}. */
+  public static Sink sink() {
+    return new AutoValue_ThriftIO_Sink.Builder().build();
+  }
+
+  /** Implementation of {@link #readFiles}. */
+  @AutoValue
+  public abstract static class ReadFiles
+      extends PTransform<PCollection<FileIO.ReadableFile>, PCollection<Document>> {
+
+    abstract Builder toBuilder();
+
+    @Override
+    public PCollection<Document> expand(PCollection<FileIO.ReadableFile> input) {
+      return input.apply(ParDo.of(new ReadFn()));
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract ReadFiles build();
+    }
+
+    /** Reads a {@link FileIO.ReadableFile} and outputs a Thrift {@link Document}. */
+    static class ReadFn extends DoFn<FileIO.ReadableFile, Document> {
+
+      @ProcessElement
+      public void processElement(ProcessContext processContext) {
+        FileIO.ReadableFile file = processContext.element();
+        try {
+          Document document =
+              ThriftIdlParser.parseThriftIdl(
+                  ByteSource.wrap(file.readFullyAsBytes()).asCharSource(Charsets.UTF_8));
```

Comment:
I think since we will only have access to the underlying file as a `byte[]` or `String` I don't know if we will be able to safely determine `Charset` at runtime, thoughts?

Reply:
Does Thrift allow characters outside of UTF-8 in IDL files?

Reply:
It doesn't look like it, I tried using 样品 and £ in a test .thrift and it failed to compile.